File: pkg/repository/post.go
package repository

import (
	"github.com/approvers/qip/pkg/domain"
	"github.com/approvers/qip/pkg/utils/id"
)

type PostRepository interface {
	Create(p domain.Post) error
	FindByID(id id.SnowFlakeID) (*domain.Post, error)
	FindByAuthorID(id id.SnowFlakeID) ([]domain.Post, error)
	FindByIDWithUserIcon(id id.SnowFlakeID) (*PostUserFileJoinedData, error)
	FindByAuthorIDWithUserIcon(id id.SnowFlakeID) ([]PostUserFileJoinedData, error)
}

type PostUserFileJoinedData struct {
	Post *domain.Post
	User *domain.User
	File *domain.File
}

File: src/repository/post.ts
import { AsyncResult, Result } from "../helpers/result.js";
import { Post } from "../domain/post.js";
import { Snowflake } from "../helpers/id_generator.js";
import { User } from "../domain/user.js";

export interface IPostRepository {
  Create(p: Post): Promise<Result<Post, Error>>;
  Update(p: Post): Promise<Result<Post, Error>>;

  FindByID(id: Snowflake): Promise<Result<Post, Error>>;
  FindByAuthor(id: Snowflake): Promise<Result<Array<Post>, Error>>;

  ChronologicalPosts(
    authorIDs: Snowflake,
    cursor: number,
  ): AsyncResult<{ posts: Post; author: User }[], Error>;

  Delete(id: Snowflake): AsyncResult<void, Error>;
}
